ocaml/typing/ctype.ml

```diff
@@ -2091,6 +2110,55 @@ let tvariant_not_immediate row =
       | _ -> false)
     (row_fields row)
 
+
+(* Jkinds *)
+
+let is_datatype_decl (k : type_decl_kind) =
+  match k with
+  | Type_record _ | Type_variant _ | Type_open -> true
+  | Type_abstract _ -> false
+
+let find_type_opt p env =
+  try Some (Env.find_type p env) with Not_found -> None
+
+let rec jkind_of_decl_unapplied env (decl : type_declaration) : jkind option =
+  (* FIXME jbachurski: Shouldn't we look at type_variance and type_separability here? *)
+  match decl.type_arity with
+  | 0 -> Some decl.type_jkind
+  | _ when is_datatype_decl decl.type_kind ->
+    Some (Arrow
+    { args = List.map (type_jkind env) decl.type_params
+    ; result = decl.type_jkind })
+  | _ -> None
+
+and type_jkind_for_app (app_jkind : jkind) app_arity tys : jkind option =
+  match tys, app_jkind with
+  | Unapplied, _ when app_arity = 0 -> Some app_jkind
+  | Unapplied, _ -> None
+  | Applied tys, _ when app_arity > 0 ->
+    if List.length tys = app_arity
```

Comment:
It's probably fine to just assume the arities are correct here rather than checking them.

Reply:
I think I see why it'd be okay to drop the check, but it'd make me more uncomfortable if it wasn't there. Is there a reason to drop it beside performance?